File: data_structure/Graph/graph_order.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>
#include <iterator>
#include <algorithm>
using namespace std;

vector<vector<int>> adj;

void makeGraph(const vector<string>& words){
	adj = vector<vector<int>>(26, vector<int>(26, 0));
	for(int j=1; j < words.size(); ++j){
		int i = j-1; int len = min(words[i].size(), words[j].size());

		for(int k=0; k < len; ++k){
			if(words[i][k] != words[j][k]){
			int a = char(words[i][k]) - 'a';
			int b = char(words[j][k]) - 'a';
			adj[a][b] = 1;
			break;
			}
		}
	}
}

int main(){

	unordered_set<char> answer;
	vector<string> words = {"google", "apple", "microsoft"};
	makeGraph(words);

	for(int i=0; adj.size(); ++i){
		for(int j=0; adj[i].size(); ++j){
			if(adj[i][j] == 1)
				answer.insert(char(i + 97));
		}
	}

	for(unordered_set<char>::const_iterator it = answer.cbegin(); it != answer.cend(); ++it){
		cout << *it << " ";
	} cout << endl;

	return 0;
}

File: Python/graph_order.py

adj = [[0] * 26] * 26 # Store adjacent matrix graph in 26x26 list

# makeGraph will find arbitrary dictionary order from the given string list
# For instance, order(make -> cake) in the strings, 'm' should come first than 'c'
def makeGraph(words):

	for i in range(1, len(words)):
		j = i-1 # compare words[i] and words[j]
		min_len = min(len(words[i]), len(words[j])) # return min search length
		for k in range(min_len):
			if words[i][k] != words[j][k]:
				a = ord(words[i][k]) - ord('a') #ord will return the int value of ASCII character
				b = ord(words[i][k]) - ord('a')
				adj[a][b] = 1	# If a -> b is order, convert matrix value to one


# create examples string list and empty set
examples = ["microsoft", "apple", "google"]
makeGraph(examples)
answers = set()	
	
for i in range(len(adj)):
	for j in range(len(adj[i])):
		if adj[i][j] == 1:
			answers.add(str(chr(i + 97))) #add ASCII converted character to the answers set
# print result
print(answers)